feat: Adding Caching CMM

seebees opened the pull request:
Adding a Caching CMM implementation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Resolves https://github.com/aws/aws-cryptographic-material-providers-library/issues/480

texastony commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy:
```diff
@@ -0,0 +1,1100 @@
+                     && PutCacheEntryInput.identifier == Seq.Last(cache.History.GetCacheEntry).input.identifier
+                     && PutCacheEntryInput.bytesUsed == input.maxPlaintextLength
+                 )
+                 //= aws-encryption-sdk-specification/framework/caching-cmm.md#get-encryption-materials
+                 //= type=implication
+                 //# If the [algorithm suite](algorithm-suites.md) on the returned encryption material
+                 //# contains an an [Identity KDF](algorithm-suites.md#identity-kdf),
+                 //# or the newly constructed [cache entry is not within limits](#cache-entry-within-limits),
+                 //# the caching CMM MUST NOT store the encryption materials in the underlying CMC.
+                 && (
+                   || encryptionMaterials.algorithmSuite.kdf.IDENTITY?
+                   || (exists now
+                         :: !cacheEntryWithinLimits(
+                              creationTime := now,
+                              bytesUsed    := input.maxPlaintextLength.value,
+                              messagesUsed := 1,
+                              now          := now
+                            ))
+                   ==>
+                     && |cache.History.PutCacheEntry| == |old(cache.History.PutCacheEntry)|
+                 )
```
Note: I've always wondered about how KDF is used for the Caching CMM.
What I don't understand is when the KDF is invoked.
i.e: We have returned and cached materials b/c a KDF is in the Algorithm Suite.
But when are we are going to KDF?

Only during content encryption?

seebees replied:
Yes, during content encryption.
The materials are not "smart" so they leave the do the KDF to the caller.

texastony commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy:
```diff
@@ -0,0 +1,1100 @@
+        && |cryptoPrimitives.History.Digest| == |old(cryptoPrimitives.History.Digest)| + |input.encryptedDataKeys| + 2
+        && var outputRequest := Seq.Last(cryptoPrimitives.History.Digest);
+        && var edkDigestRequests
+          := cryptoPrimitives.History.Digest[
+             |old(cryptoPrimitives.History.Digest)| .. |old(cryptoPrimitives.History.Digest)| + |input.encryptedDataKeys|
+             ];
```
Question: Is this selection by index of History of Digest calls?
I'm not familiar with `..` in Dafny,
but I presume it is slicing.

I am confused because `|old(cryptoPrimitives.History.Digest)|` is used twice...
OH, select from `oldIndex` to `oldIndex` + EDK count. 
Got it.

seebees replied:
Yes, it is slice.
Since there are some number of EDKs, I need to get all of them.
http://dafny.org/dafny/DafnyRef/DafnyRef#sec-other-sequence-expressions

texastony reviewed (dismissed):
I know it's not finished, but I don't have any blocking complaints.

seebees commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy:
```diff
@@ -153,11 +153,66 @@ operation CreateRequiredEncryptionContextCMM {
+  @javadoc("The Cryptographic Materials Cache the Caching Cryptographic Materials Manager will use to cache requests.")
+  underlyingCMC: CryptographicMaterialsCacheReference,
+  @required
+  @javadoc("Sets the maximum lifetime for entries in the cache, for both encrypt and decrypt operations. When the specified amount of time passes after initial creation of the entry, the entry will be considered unusable, and the next operation will incur a cache miss.")
```
Hum, I think I want operations but I'm not sure.
The point here is that there is 1 cache.
But that items put in from encrypt are not hits on decrypt.

seebees commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy:
```diff
@@ -153,11 +153,66 @@ operation CreateRequiredEncryptionContextCMM {
+  //# - [Partition ID](#partition-id)
+  //# - [Limit Bytes](#limit-bytes)
+  //# - [Limit Messages](#limit-messages)
+  @javadoc("Sets the partition ID for this CMM. By default, two CMMs will never use each other's cache entries. This helps ensure that CMMs with different delegates won't incorrectly use each other's encrypt and decrypt results. However, in certain special circumstances it can be useful to share entries between different CMMs - for example, if the backing CMM is constructed based on some paramete ...
```
This content comes from the ESDK-Java.
So it is somewhat public.
I'd like to get `javadoc` to work in every language so we don't have to worry about this :)

seebees commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy:
```diff
@@ -0,0 +1,1100 @@
+          && input == Seq.Last(underlyingCMM.History.GetEncryptionMaterials).input
+          && output == Seq.Last(underlyingCMM.History.GetEncryptionMaterials).output
+
+      ensures
+        && !BypassCache(input, limitBytes)
+        && output.Success?
+        ==>
+          && |cryptoPrimitives.History.Digest| == |old(cryptoPrimitives.History.Digest)| + 2
+          && var ecDigestRequest := Seq.Last(Seq.DropLast(cryptoPrimitives.History.Digest));
+          && var cacheEntryIdentifierRequest := Seq.Last(cryptoPrimitives.History.Digest);
+          && ecDigestRequest.output.Success?
+          && cacheEntryIdentifierRequest.output.Success?
+          && CanonicalEncryptionContext.EncryptionContextToAAD(input.encryptionContext).Success?
+          && ecDigestRequest.input
```
Do you mean something like
// These values are setup for use later in the proof

In order to prove the various branches these values are need.

seebees commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy:
```diff
@@ -0,0 +1,1100 @@
+    0,0,0,0,0,0,0,0
+  ]
+
+  lemma PaddingIsCorrect()
```
I was very pleased myself.
Especially since the first time I wrote it I have 8x9!

seebees commented on AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy:
```diff
@@ -0,0 +1,1100 @@
+    output := cache.PutCacheEntry(input);
+  }
+
+  method EncryptMaterialsCacheKey (
```
Name changed

seebees committed:
feat: Adding Caching CMM

Adding a Caching CMM implementation

Add operation and work around parsing error

Add the CachingCMM operation

adding duvet

no unbounded int!

Updates to CachingCMM

Small update to the create operation

small update

Update AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy

Co-authored-by: Tony Knapp <5892063+texastony@users.noreply.github.com>

Update AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy

Co-authored-by: Tony Knapp <5892063+texastony@users.noreply.github.com>

Update AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy

Co-authored-by: Tony Knapp <5892063+texastony@users.noreply.github.com>

Update AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cmms.smithy

Co-authored-by: Tony Knapp <5892063+texastony@users.noreply.github.com>

better names

codegen Java

Updates and format

dafny poly

seebees committed:
Small updates

seebees committed:
first cut

josecorella committed:
Merge branch 'main' into ryanemer/caching-cmm

josecorella committed:
Merge branch 'main' into ryanemer/caching-cmm

josecorella committed:
regenerate net code

josecorella committed:
regenerate java code

josecorella committed:
regenerate python code

josecorella committed:
regenerate go code

josecorella committed:
format

josecorella committed:
Merge branch 'main' into ryanemer/caching-cmm
